### PR TITLE
EventDB improvements 

### PR DESCRIPTION
**Pull request by mapaaa**
*Issue #, if available:*

*Description of changes:*

EventDB now keeps events in separate vectors by the klass id. Each vector is sorted in ascending order by timestamps. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/viewer/test_event_db.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+using namespace HawkTracer::viewer;
+using HawkTracer::parser::Event;
+using HawkTracer::parser::EventKlass;
+using HawkTracer::parser::EventKlassField;
+using HawkTracer::parser::FieldTypeId;
+using HawkTracer::parser::FieldType;
+
+class TestEventDB : public ::testing::Test
+{
+    protected:
```
Comment (loganek) [inline tests/viewer/test_event_db.cpp]: could you please remove the indentation?

File: tests/viewer/test_event_db.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    FieldType _value3{};
+    FieldType _value4{};
+    EventDB event_db;
+};
+
+TEST_F(TestEventDB, CheckIfEventsAreKeptSorted)
+{
+    // Arrange
+    Query query;
+    HT_TimestampNs start_ts = (HT_TimestampNs)100;
```
[the remark this answers is not in the document]
Comment (loganek) [reply tests/viewer/test_event_db.cpp]: The same for other tests

File: tests/viewer/test_event_db.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    // Act
+    event_db.insert(*_event1);
+    event_db.insert(*_event2);
+    event_db.insert(*_event3);
+    auto response = event_db.get_data(start_ts, stop_ts, query);
+
+    // Assert
+    std::vector<Event> correct_response = {*_event1, *_event3};
+    ASSERT_EQ(response.size(), correct_response.size());
+    for (unsigned int i = 0; i < response.size(); ++i)
```
Comment (loganek) [inline tests/viewer/test_event_db.cpp]: `size()` returns value of type `size_t`, so `i` should also be of `size_t`

File: tests/viewer/test_event_db.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    HT_TimestampNs start_ts;
+    HT_TimestampNs stop_ts;
+
+    // Act
+    event_db.insert(*_event1);
+    event_db.insert(*_event2);
+    event_db.insert(*_event3);
+    event_db.insert(*_event4);
+
+    start_ts = (HT_TimestampNs)100;
```
Comment (loganek) [inline tests/viewer/test_event_db.cpp]: you could use `u` suffix, so explicit cast to `HT_TimestampNs` won't be necessary; i.e. `start_ts = 100u;`

File: tests/viewer/test_event_db.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        ASSERT_EQ(response2[i].get().get_timestamp(), correct_response2[i].get_timestamp());
+    }
+    ASSERT_EQ(response3.size(), correct_response3.size());
+    for (unsigned int i = 0; i < response3.size(); ++i)
+    {
+        ASSERT_EQ(response3[i].get().get_timestamp(), correct_response3[i].get_timestamp());
+    }
+}
+
```
Comment (loganek) [inline tests/viewer/test_event_db.cpp]: I think we need to have a few more tests:
 * Could you add test where `start_ts > stop_ts`?
 * Could you add test where database is empty?
 * Could you add test where `start_ts` == `stop_ts` and one of the events' timestamp is equal to `start_ts`?

File: viewer/event_db.cpp [hunk 6]
```diff
@@ -1,38 +1,122 @@
...
+    if (_events.find(klass_id) == _events.end())
+    {
+        _events[klass_id] = std::vector<parser::Event>(1, std::move(event));
+    }
+    else
+    {
+        std::vector<parser::Event>::reverse_iterator it;
+        for (it = _events[klass_id].rbegin(); it != _events[klass_id].rend(); ++it)
+        {   
+            if ((*it).get_timestamp() < event.get_timestamp())
```
Comment (loganek) [inline viewer/event_db.cpp]: You can get rid of this explicit dereference; instead of `(*it).get_timestamp()`, you can write `it->get_timestamp()`

File: viewer/event_db.cpp [hunk 7]
```diff
@@ -1,38 +1,122 @@
...
 {
-    std::vector<EventRef> events;
+    std::vector<EventRef> response;
+
+    // Dummy events for calling lower_bound
+    DummyEvent dummy_start_event(start_ts);
+    DummyEvent dummy_stop_event(stop_ts);
 
-    for (const auto& event : _events)
+    if (query.klass_id == (HT_EventKlassId)-1)
```
Comment (loganek) [inline viewer/event_db.cpp]: could we move those two cases (`klass_id == -1` and `klass_id != -1`) to two separated functions? The code would be easier to follow

File: viewer/event_db.cpp [hunk 8]
```diff
@@ -1,38 +1,122 @@
...
+            const auto first_event = std::lower_bound(events.second.begin(), events.second.end(), dummy_start_event,
+                    [] (const parser::Event& e1, const parser::Event& e2) {
+                    return e1.get_timestamp() < e2.get_timestamp();
+                    });
+            const auto last_event = std::lower_bound(events.second.begin(), events.second.end(), dummy_stop_event,
+                    [] (const parser::Event& e1, const parser::Event& e2) {
+                    return e1.get_timestamp() <= e2.get_timestamp();
+                    });
+            ++klass_index;
+            for (auto it = first_event; it != last_event; ++it)
```
Comment (loganek) [inline viewer/event_db.cpp]: looks like you have the same for loop in the `else` branch; could you move this code to separated function, instead of duplicating it?

File: viewer/event_db.cpp [hunk 9]
```diff
@@ -1,38 +1,122 @@
...
 
-    for (const auto& event : _events)
+    if (query.klass_id == (HT_EventKlassId)-1)
     {
-        auto ts = event.get_timestamp();
-        bool get_id_ok = query.klass_id == (HT_EventKlassId)-1 || event.get_klass()->get_id() == query.klass_id;
-        bool ts_ok = ts >= start_ts && ts <= stop_ts;
+        // Query all the event klasses
+        std::vector<int> lengths(_events.size(), 0);
+        int klass_index = -1;
```
Comment (loganek) [inline viewer/event_db.cpp]: if you use the variable as an index of an array, it shouldn't be `signed int`, as it doesn't make sense to have negative index. Usually it should be `size_t`. However, I don't think you need this function at all, you could just insert values to `lengths` using `push_back` method. If you want to pre-allocate memory in the vector, you can call `vector::reserve()` method.

File: viewer/event_db.cpp [hunk 10]
```diff
@@ -1,38 +1,122 @@
...
+        for (const auto& events : _events)
```
Comment (loganek) [inline viewer/event_db.cpp]: Body of this loop is very similar to a code in `else` branch. Could we somehow move this code to a function, so we have the implementation only once?

File: viewer/event_db.hpp [hunk 11]
```diff
@@ -26,10 +26,34 @@ class EventDB
...
+    class CompareTimestampEvent : public parser::Event
+    {
+    public:
+        CompareTimestampEvent(HT_TimestampNs timestamp)
+        {
+            _timestamp = timestamp;
+        }
+    };
+
+    void query_all_event_klasses(std::vector<EventRef>& response,
```
Comment (loganek) [inline viewer/event_db.hpp]: we use `_` prefix for private/protected data. Could you do that?
